### DISCUSSION TITLE
async traits

### DISCUSSION THREAD

**Discussion by junderw** [rust-bitcoin/rust-bitcoin#2407] [General]
Starting a discussion on async traits.

Two topics of discussion I'd like to start with:

1. Is the `async_trait` crate acceptable or should we wait until we bump MSRV to >= 1.75?
2. What are some APIs that would benefit from offering an async version of the trait?

I think for 1 that waiting a few years for MSRV to get high enough is probably desirable.

For 2, I think signing related APIs tend to have use cases for async/await since hardware wallets etc. wait on user input and wallet wait on the hardware wallets.

**Comment by Kixunil** [comment]
I think async traits are entirely unnecessary in the `bitcoin` crate. We're not doing IO except for serialization/deserialization, which will be abstracted over sync/`async` ⨯ `std`/`no_std` using `push_decode` - see #2184

The only one which may need this is `bitcoincore-rpc` but I'm not sure if async traits are needed there either.

**Comment by junderw** [comment]
> We're not doing IO

But the wallet *using* the bitcoin crate might.

What if someone has a wallet that needs to send the data over a socket to some HSM?

They need to wait for the HSM response in that case. This is IO.

In BitcoinJS we handle this use case by accepting a `Signer` argument for `signInput` and a `SignerAsync` argument for `signInputAsync` etc. This allows for those people to implement some class that implements the SignerAsync interface and pass it in. We don't care about the implementation.

Something like that would be useful in Rust as well. An abstraction over signing, which PrivateKey and Xpriv would implement. Then, alternatively, an async version that people could choose to implement for their IO based signing setup...

Not sure if these kinds of abstractions should be higher up in something like bdk, but since a lot of HSMs are pretty low level and only concern themselves with signing hashes/preimages, it seems like the abstraction should be pretty low.

**Reply by Kixunil** [reply]
> But the wallet _using_ the bitcoin crate might.

Doesn't matter, `async` code can call non-blocking sync code and all our sync code is non-blocking.

> What if someone has a wallet that needs to send the data over a socket to some HSM?

They just split the calls like all other Rust code does:

```rust
let foo = bitcoin::something();
let bar = hsm.call_with_foo(foo).await;
let baz = bitcoin::use_bar(bar);
```

There's no problem with that.

> An abstraction over signing, which PrivateKey and Xpriv would implement.

That really shouldn't be required, the crate already provides sufficiently low-level primitives to not need that and the `Sign` trait proposed in #2405 is just for a convenience method. Just write two more lines, no big deal. And if you want to avoid those two lines, just define your own trait and one day when MSRV is fine, it can be added. Note that I really don't want to abstract over `async` unless there's full language support, so it'd be just a new trait and new methods.

Anyway I've made an issue for you to make sure nothing slips through: https://github.com/rust-bitcoin/rust-bitcoin/issues/2409

**Comment by dpc** [comment]
Async is an unfinished Rust flavor that is a PITA to use and I'd advise everyone to avoid it as much as they can. [If you're interested about what I think about it: you can take a look](https://lobste.rs/s/jkct2m/avoid_async_rust#c_qingfz). There is no way in Rust to be generic over async, so you'll inevitability having to deal with worst of both worlds if you want to support both, which means you'll just support async like most of the ecosystem does already: https://www.reddit.com/r/rust/comments/197811x/the_bane_of_my_existence_supporting_both_async/

IMO rust-bitcoin should avoid it as much as possible by being IO-agnostic. If software needs to read and deserialize in an async context, it should just read into memory first, then deserialize later, or figure out something else (push based encoding). Maybe there's a room for separate crate that does consensus encoding on async io traits (are these even stable thing yet?).

**Reply by Kixunil** [reply]
> Async is an unfinished Rust flavor that is a PITA to use and I'd advise everyone to avoid it as much as they can.

It is PITA but avoiding it doesn't seem like a good solution.

> so you'll inevitability having to deal with worst of both worlds if you want to support both

Not if the only thing we need to do is encoding and decoding - [`push_decode`](https://docs.rs/push_decode) is a solution (well, it does blow up the code until generators are stable and full-featured but it can be managed with macros).

> IMO rust-bitcoin should avoid it as much as possible by being IO-agnostic.

That's pretty much what `push_decode` does. And it is also agnostic over `std`/`no_std` and over IO error types (there actually aren't any, just a bridging enum) so we can have panic-free hashing. (Finally!)

> If software needs to read and deserialize in an async context, it should just read into memory first, then deserialize later

That'd be simplest and maybe also fastest but you can't do that with bitcoin's consensus encoding because you don't know the size of data to read upfront.

> Maybe there's a room for separate crate that does consensus encoding on async io traits

No need to do that because bridging between `push_decode` and `tokio`/`async_std` is around 4 LOC.

> are these even stable thing yet?

Those in `tokio` and those in `async-std` are stable with respect to those crates and they are not shared, so if `std` adds them it'll break the ecosystem. :angry: But at least they are stable enough that I'd agree to have them in `bitcoin-consensus-encoding` 1.0.

**Reply by dr-orlovsky** [reply]
> Async is an unfinished Rust flavor that is a PITA to use and I'd advise everyone to avoid it as much as they can.

Can't agree more! Instead of async it is much better to use the reactor pattern (that's why @cloudhead and I did [`io-reactor` crate](https://crates.io/crates/io-reactor)).

The only problem is WASM, where there is no other option than async.

**Reply by dpc** [reply]
WASM is another layer of PITA.

> That'd be simplest and maybe also fastest but you can't do that with bitcoin's consensus encoding because you don't know the size of data to read upfront.

Just doing blocking IO (possibly offloaded from async executor via `block_in_place`/`spawn_blocking` is the way to go.

**Reply by Kixunil** [reply]
Not entirely on topic but I think there's an interesting special case with LN and similar protocols where the number of connections is bounded by money: you can write a simple `async` connection acceptor that only handles handshakes, gossip and opening channels and then have a thread/process for each channel. The rationale is that it's better to have `async` IO for many connections but for few connections with not too frequent traffic blocking is OK. This assumes channel handling is much more complex than the rest so that the complexity of having both approaches is justified.

@dr-orlovsky I still don't understand `io-reactor`, I think an example of an "echo" server would be helpful.

Note that there's one more thing I believe `async` does better than blocking: task cancellation. I know of only two ways to do that in blocking. First is to use a signal, make sure the signal mask doesn't block it when calling IO and handling `Interrupted` which is often silently ignored in things like `read_exact`, `io::copy` etc so you have to somehow ban calling them and write your own alternatives and document in user documentation that the signal is internal and users should not send it.

Second is to simulate blocking by blocking on `poll` with each operation, after which the thread checks whether cancellation was requested and then calls the real operation in non-blocking way. This whole thing has to be hardwired into the task.

With `async` you just wrap the future in some cancellation helper that internally has a one shot channel and returns `Err(Cancelled)` when you cancel it. This composes nicely as long as you don't require a future returning a specific error type somewhere. (And if you do, just refactor it with the help of the compiler.)

> WASM is another layer of PITA.

Not really that much, IME. The only thing that's really annoying is that you can't rely on it not having threads because AFAIK you can't prevent the code from compiling on a hypothetical new version of wasm which does have threads.

**Reply by dpc** [reply]
> @dr-orlovsky I still don't understand `io-reactor`, I think an example of an "echo" server would be helpful.

Similar perception here. I kind of understand what it is aiming at, but without more elaborated examples it was hard for me to judge how how well it would work in practice.

>  The rationale is that it's better to have async IO for many connections but for few connections with not too frequent traffic blocking is OK. 

Applications that can should wrap Bitcoin Consensus encodings into more robust encoding. Bitcoin Consensus encoding is optimized for compactness of serialization, and sacrifices robustness of deserialization compared to TLV and similiar schemes. But it doesn't cost much to "wrap" consensus encoded things into a "lenght-prefix" and thus be able to read them first as bytes, and deserialize without doing any IO afterwards.

Worth pointing out that 64 bit Linux will easily handle 10k threads handling 10k connections, and that's a lot of concurrent connections, even for a server. For a Bitcoin application, is it really worth it introducing all the delicate and complex async reactor machinery which will inevitably increase a chance of CVEs and multiply technical difficulty, just to make it all use slightly less CPU?

> I believe async does better than blocking: task cancellation

Indeed, it's a really nice property of async, when one needs it. However it goes both ways as now there's a huge surface of introducing cancellation unsafety bugs, as at any `await` your code can stop executing out of the blue. It's not all that bad, but just another axis to watch out for at all times, and in bigger teams developers are not used to pay attention to it.

In blocking code it's actually rarely really needed to do cancellations. Software with persisted state should be crash-safe, so already needs to be able to tolerated hard-kill if someone is in a rush. And soft kill can be just a top-level flag that tasks can check periodically and shut-down cleanly, terminating gracefully instead of mid-write/read.

Even in Fedimint (again: async ccode) we use a top-level task cancellation abstraction, so that on termination request all tasks wrap up what they were doing cleanly, not have to spam logs with random errors. Async timeouts are used tactically for things that race with each other and outgoing requests. Both could be somewhat easily done with a combination of channels and IO deadlines (IO timeouts). BTW. deadline abstraction was never properly developed into Rust stdlib unfortunately.

> > WASM is another layer of PITA.
> 
> Not really that much, IME.

Conditional `: MaybeSend + MaybeSync` and `#[apply(maybe_send_async_trait)]` magic for wasm in Fedimin it's annoying, and is yet another axis of complexity. And the problem with complexity like this is that it adds exponentially. So now you worry about limitations of the normal language * lifetimes * ownership * limitations of async * workarounds for WASM * your business issues.

**Reply by Kixunil** [reply]
> But it doesn't cost much to "wrap" consensus encoded things into a "lenght-prefix" and thus be able to read them first as bytes, and deserialize without doing any IO afterwards.

It's not that much (just serialize twice, first time to a sink) but you are incompatible with whatever protocol chose to not do that (outside Bitcoin: HTTP has the same problem. :D)

> introducing all the delicate and complex async reactor machinery which will inevitably increase a chance of CVEs

Good point.

> cancellation unsafety bugs

IIRC I never had a problem with this, as you point out it's not that bad.

But also I think the `Future` trait has a few shortcomings that unfortunately I was unable to explain and solve when it was being designed (I did watch the discussion and tried to help at the time.) I think the trait should've had `unsafe` methods, parametrized context and maybe parametrized `Poll`. And probably other things I don't remember now. So part of the problem is "current `async` Rust" vs "theoretical `async` Rust".

> In blocking code it's actually rarely really needed to do cancellations.

I suspect this strongly depends on the kind of application. For instance cancellation is frequent and important in GUI applications where the user should be able to cancel an operation (e.g. something as simple as downloading a file in browser). In other applications probably not so much.

> should be crash-safe

I guess 99.99% of software just isn't. It takes a crazy effort to do that. E.g. this is a crash safe version of what's effectively `x += 1`: https://github.com/debian-cryptoanarchy/cryptoanarchy-deb-repo-builder/blob/master/ridetheln/alloc_index.sh
I try to write my own software to be crash-safe but it's because I'm crazy.

>  tasks can check periodically and shut-down cleanly

Well, that's the problem, you really can't. If your browser is stuck waiting for data to arrive the syscall just waits "forever" (5 min). And if the user hits "cancel" and it doesn't respond the application looks unresponsive - i.e. crappy.

> we use a top-level task cancellation abstraction, so that on termination request all tasks wrap up what they were doing cleanly, not have to spam logs with random errors.

Cool!

> Conditional `: MaybeSend + MaybeSync` and `#[apply(maybe_send_async_trait)]` magic for wasm in Fedimin it's annoying

Didn't see that, I will try to take a look later.

**Reply by dr-orlovsky** [reply]
@Kixunil, @dpc:

> @dr-orlovsky I still don't understand io-reactor, I think an example of an "echo" server would be helpful.

> Similar perception here. I kind of understand what it is aiming at, but without more elaborated examples it was hard for me to judge how how well it would work in practice.

So, the idea is simple: we split business logic (service) from I/O. Reactor is a thing that manages I/O event loop: first, it locks (epoll) on _ALL_ I/O sources (net, files etc) and when data are available, runs the service for each data source. The service is not provided with anything which can lock, if it needs to write/output, it tells reactor - and reactor ensures that only  non-blocking write is made (otherwise we buffer). Everything long-running is delegated to worker threads via crossbeam channels (again, no async, arc, mutexes and locks).

More elaborated low-level description can be found in https://cloudhead.io/popol/

You can also read the wiki page about the reactor pattern: https://en.wikipedia.org/wiki/Reactor_pattern

Today, `io-reactor` powers Radicle (decentralized github w/o blockchain :) - https://radicle.xyz

I am currently working on building append-only DB (for RGB & blockchain indexer) and WebSockets server which will use `io-reactor`.

**Reply by Kixunil** [reply]
I meant I'd like to understand usage, not necessarily how it works.

BTW, splitting business logic from IO is a good idea anyway because it makes testing much easier. I've known this for a while. But usually it's accomplished by deserializing first and passing deserialized objects to functions, possibly via channels.

**Reply by dr-orlovsky** [reply]
> I meant I'd like to understand usage, not necessarily how it works.

The usage is where you need to work with a lot of blocking resources in a much more efficient way than `async` allows: with the reactor you can serve tens of thousands of network connections over one I/O thread, and avoid plenty of boilerplate code since you do not need `Arc/Mutext`, `await?` etc. - thus, you also get rid of many errors which you have to handle in the business logic.

**Reply by dpc** [reply]
Do I read it right, that's it's basically streamlined `mio`?

**Reply by Kixunil** [reply]
@dr-orlovsky I meant a piece of simple code that uses it as a demo. Like an echo server.

**Reply by dr-orlovsky** [reply]
I think https://github.com/cyphernet-dao/nsh can be seen as a simple sample project using `io-reactor`. Specifically, you can look here: https://github.com/cyphernet-dao/nsh/blob/master/src/server.rs

It is an analog of SSH tool, but (a) using Noise framework instead of OpenSSH for the encryption and (b) using `io-reactor`

PS. Here is the `echo` command you was asking about: https://github.com/cyphernet-dao/nsh/blob/836323230693bbad4fc3d508c12d35f7e8c1b7e6/src/processor.rs#L76-L93

**Reply by Kixunil** [reply]
Cool, I think I understand it now. Two things jumped at me:

* There's quite a bit of boilerplate because all methods of `Handler` are mandatory. I think the common case is just IO handling, not timers.
* You pass `RawFd` in public API which clearly excludes Windows support (I wouldn't need it but some people might).

**Reply by dr-orlovsky** [reply]
Regarding (1) I think you are right and there can be a default implementation doing nothing.

Regarding the (2) you are not oblidged to use `RawFd`; it is just the approach taken by the demo

But yes, reactor is not yet windows-compatible, that is one of the tasks I plan to address in some foreseable future.

**Comment by dpc** [comment]
> Well, that's the problem, you really can't. If your browser is stuck waiting for data to arrive the syscall just waits "forever" (5 min).

Whatever async IO is using under the hood, can be done in blocking IO code to enforce a deadline on a timer / select on a flag. It would have been slightly more "manual", but Rust is very good at building abstractions. And even if slightly cumbersome, it would only needed to be used where actually needed, instead of turning the language inside out, breaking composability of the ecosystem and infecting even code that doesn't care about it with all the async complexity.

And ... if someone really needs it, they should only do *that* in async code, in a limited scope. I don't dislike async as much as I dislike how prevalent it became. For decades software of all kinds was built without async and suddenly it's ["async, async, async"](https://www.youtube.com/watch?v=ycKNt0MhTkk) everywhere.

**Reply by Kixunil** [reply]
Well, maybe if the ecosystem is `async` anyway it's better to use it then reinvent cancellation. I don't hate `async` either, I think most of its problems are just language shortcomings (like lack of GAT).

**Reply by dpc** [reply]
Exactly. There's bunch of problems, though lots of them theoretically fixable. But it might take years, while the community happily jumped on the bandwagon 2 years ago.

**Comment by vincenzopalazzo** [comment]
idk if we can achieve something without implementing a PoC of this, it is possible to implement the async trait and expose them with a feature flag anyway.

The async in rust will be out iirc on 2027 for real usage, there are too many feature that are missing like async drop and `dyn*`

**Reply by Kixunil** [reply]
As was discussed this crate doesn't really need async traits, we can workaround it nicely. However, it'd really help if we had `Future<Context>` which I've seen proposed somewhere (there was an article on reddit linking to it recently), since then I could write a nice hack for `push_decode`. (Although I would have to figure out what to do about pinning which is another can of worms.)
